README.md
[Nginx-Clojure](http://nginx-clojure.github.io)
=============

[![Build Status](https://travis-ci.com/nginx-clojure/nginx-clojure.svg?branch=master)](https://app.travis-ci.com/github/nginx-clojure/nginx-clojure)
[![Clojars Project](https://img.shields.io/clojars/v/nginx-clojure.svg)](https://clojars.org/nginx-clojure)
[![BSD licensed](https://img.shields.io/badge/license-BSD-blue.svg)](./LICENSE)
![GitHub last commit](https://img.shields.io/github/last-commit/nginx-clojure/nginx-clojure)
[![SourceForge](https://img.shields.io/sourceforge/dt/nginx-clojure)](https://sourceforge.net/projects/nginx-clojure/files/)


![Alt text](logo.png) [Nginx-Clojure](http://nginx-clojure.github.io) is a [Nginx](http://nginx.org/) module for embedding Clojure or Java or Groovy programs, typically those [Ring](https://github.com/ring-clojure/ring/blob/master/SPEC) based handlers.


Core Features
=================

The latest release is v0.6.0, more detail changes about it can be found from [Release History](//nginx-clojure.github.io/downloads.html).

1. Compatible with [Ring](https://github.com/ring-clojure/ring/blob/master/SPEC) and obviously supports those Ring based frameworks, such as Compojure etc.
1. Http Services by  using Clojure / Java / Groovy to write simple handlers for http services.
1. Nginx Access Handler by Clojure / Java / Groovy
1. Nginx Header Filter by Clojure / Java / Groovy
1. Nginx Body Filter by Clojure / Java / Groovy
1. Nginx Log Handler by Clojure / Java / Groovy
1. HTTP V2 support in both standard edition and embedded edition which are compiled against Nginx 1.18.0+
1. Support Java 9, 10, 11, 12, 19
1. Support to use jdk19 built-in coroutine viz. Continuation
1. Pub/Sub Among Nginx Worker Processes
1. Shared Map based on shared memory & Shared Map based Ring session store
1. Support Sente, see [this PR](https://github.com/ptaoussanis/sente/pull/160)
1. Support Per-message Compression Extensions (PMCEs) for WebSocket
1. APIs for Embedding Nginx-Clojure into a Standard Clojure/Java/Groovy App
1. Server Side Websocket
1. A build-in Jersey container to support java standard RESTful web services (JAX-RS 2.0)
1. Tomcat 8 embedding support (so servlet 3.1/jsp/sendfile/JSR-356 websocket work within nginx!)
1. Dynamic proxying by using Clojure / Java / Groovy to write a simple nginx rewrite handler to set var or return errors before proxy pass or content ring handler
1. Non-blocking coroutine based socket which is Compatible with Java Socket API and works well with largely existing java library such as apache http client, mysql jdbc drivers. 
With this feature  one java main thread can handle thousands of connections.
1. Handle multiple sockets parallel in sub coroutines, e.g. we can invoke two remote services at the same time.
1. Asynchronous callback API of socket/Channel  for some advanced usage
1. Long Polling & Server Sent Events
1. Run initialization clojure code when nginx worker starting
1. Support user defined http request method
1. Compatible with the Nginx lastest most mainline version 1.23.3. (Nginx 1.22.X, 1.20.X, 1.18.x, 1.14.x, 1.12.x, 1.8.x, 1.6.x, 1.4.x is also ok, older version is not tested and maybe works.)
1. One of  benifits of [Nginx](http://nginx.org/) is worker processes are automatically restarted by a master process if they crash
1. Utilize lazy headers and direct memory operation between [Nginx](http://nginx.org/) and JVM to fast handle dynamic contents from Clojure or Java code.
1. Utilize [Nginx](http://nginx.org/) zero copy file sending mechanism to fast handle static contents controlled by Clojure or Java code.
1. Support Linux x64, Linux x86 32bit, Win32, Win64  and Mac OS X. Freebsd version can also be got from Freebsd ports.

By the way it is very fast, the benchmarks can be found [HERE(with wrk2)](https://github.com/ptaoussanis/clojure-web-server-benchmarks/).

Jar Repository
================

Nginx-Clojure has already been published to https://clojars.org/ whose maven repository is 

```xml
<repository>
  <id>clojars.org</id>
  <url>http://clojars.org/repo</url>
</repository>
``` 

After adding clojars repository, you can reference nginx-clojure 0.6.0 , e.g.

 Leiningen (clojure, no need to add clojars repository which is a default repository for Leiningen) 
-----------------
 
```clojure
[nginx-clojure "0.6.0"]
```
Gradle (groovy/java)
-----------------
 
```
compile "nginx-clojure:nginx-clojure:0.6.0"
```
Maven
-----------------
 
```xml
<dependency>
  <groupId>nginx-clojure</groupId>
  <artifactId>nginx-clojure</artifactId>
  <version>0.6.0</version>
</dependency>
```

Documents
=================


- [Quick Start](http://nginx-clojure.github.io/quickstart.html)
- [Downloads](http://nginx-clojure.github.io/downloads.html)
- [Installation](http://nginx-clojure.github.io/installation.html)
- [Configuration](http://nginx-clojure.github.io/configuration.html)
    - [JVM Path,Class Path & Other JVM Options](http://nginx-clojure.github.io/configuration.html#user-content-21-jvm-path--class-path--other-jvm-options)
    - [Initialization Handler for nginx worker](http://nginx-clojure.github.io/configuration.html#user-content-22-initialization-handler-for-nginx-worker)
    - [Content Ring Handler for Location](http://nginx-clojure.github.io/configuration.html#user-content-23-content-ring-handler-for-location)
    - [Coroutine/Asynchronous Client Channel/Thread Pool](http://nginx-clojure.github.io/configuration.html#user-content-24-chose--coroutine-based-socket-or-asynchronous-socketchannel-or-thread-pool-for-slow-io-operations)
- [Nginx Handlers](http://nginx-clojure.github.io/configuration.html#user-content-25-nginx-rewrite-handler)
    - [Nginx Rewrite Handler](http://nginx-clojure.github.io/configuration.html#user-content-25-nginx-rewrite-handler)
    - [Nginx Access Handler](http://nginx-clojure.github.io/configuration.html#user-content-26-nginx-access-handler)
    - [Nginx Header Filter](http://nginx-clojure.github.io/configuration.html#user-content-27-nginx-header-filter)
    - [Nginx Body Filter](http://nginx-clojure.github.io/configuration.html#user-content-28-nginx-body-filter)
    - [Nginx Log Handler](http://nginx-clojure.github.io/configuration.html#user-content-29-nginx-log-handler)
- Advanced Topic
    - [Embedding Nginx-Clojure into a standard App](http://nginx-clojure.github.io/embed.html)
    - [Server Channel for Long Polling & Server Sent Events](http://nginx-clojure.github.io/more.html#user-content-34-server-channel-for-long-polling--server-sent-events-sse)
    - [Pub/Sub Among Nginx Worker Processes](http://nginx-clojure.github.io/subpub.html)
    - [Shared Map & Session Store](http://nginx-clojure.github.io/sharedmap.html)
    - [Asynchronous Client Channel](http://nginx-clojure.github.io/more.html#user-content-36-asynchronous-client-channel)
    - [About Logging](http://nginx-clojure.github.io/more.html#user-content-37--about-logging)
    - [Sever Side WebSocket](http://nginx-clojure.github.io/more.html#user-content-38--sever-side-websocket)
    - [Java standard RESTful web services with Jersey](http://nginx-clojure.github.io/more.html#user-content-39--java-standard-restful-web-services-with-jersey)
    - [Embedding Tomcat](http://nginx-clojure.github.io/more.html#user-content-310-embeding-tomcat)
    - [Use Spring Framework with Nginx-Clojure Java Handlers](https://github.com/nginx-clojure/nginx-clojure/tree/master/example-projects/spring-core-example)
    - [Example Project about Jersey & Spring with Nginx-Clojure](https://github.com/nginx-clojure/nginx-clojure/tree/master/example-projects/jersey-spring-example)
    - [More about Nginx Worker Process](http://nginx-clojure.github.io/more.html#user-content-311-more-about-nginx-worker-process)
    - [More about Nginx-Clojure](http://nginx-clojure.github.io/more.html)
- [Directives Reference](http://nginx-clojure.github.io/directives.html)
- [API Reference (Clojure)](http://nginx-clojure.github.io/api/)
- [Example Projects](https://github.com/nginx-clojure/nginx-clojure/tree/master/example-projects)
- [Github Repository of Documents](https://github.com/nginx-clojure/nginx-clojure.github.io)


License
=================
Copyright © 2013-2023 Zhang, Yuexiang (xfeep) and released under the BSD 3-Clause license.

This program uses:
* Re-rooted ASM bytecode engineering library which is distributed under the BSD 3-Clause license
* Modified Continuations Library written by Matthias Mann  is distributed under the BSD 3-Clause license


HISTORY.md
Downloads & Release History
=============

1. [Binaries of Releases](http://sourceforge.net/projects/nginx-clojure/files/)
1. [Sources of Releases](https://github.com/nginx-clojure/nginx-clojure/releases)

## 0.6.0 (2023-03-18)

1. New Feature: #270 All handlers can be used at http & server context
1. New Feature: #272 Support Nginx 1.23.X where some internals have been changed
1. New Feature: #250 Support to use jdk19 built-in coroutine viz. Continuation
1. Binaries Distribution: Built with Nginx v1.23.3
1. Documents: Add build notes for nginx-clojure-embed

## 0.5.3 (2022-03-10)

1. Bug Fix: #256 NginxClojureAsynSocket isClosed is not return correct result
1. Binaries Distribution: Built with Nginx v1.20.2


## 0.5.2 (2020-12-23)

1. Bug Fix: #234 Try to fix no response when NGX_AGAIN return at next header filter
1. Bug Fix: #233 Fix compiler warnings when there's no zlib found
1. Enhancement: Delayed update to improve setVariable/set-ngx-var! performance at thread-pool mode
1. Example Project: Add an example project for Jersey & Spring DI
1. Example Project: Add example for integration with Spring framework
1. Binaries Distribution: Built with Nginx v1.18.0


## 0.5.1 (2019-11-23)

1. Bug Fix: Connection hangs with header filter at thread-pool mode #209 #153
1. Bug Fix: Body filter hangs when body size is larger than the value specified in proxy_buffers #219
1. Bug Fix: Segment fault caused by request is marked as closed too late #222
1. Code Style: Fix generic warnings in java code code-style #223 (Thanks to Michael @mgoblin)
1. Bug Fix: NPE caused by damaged memory because unsafe modify response headers #198 
1. Bug Fix: Wrong response for uncompressed message with PMCE enabled
1. CI: both auto-triggered unit test and integration test use travis-ci

## 0.5.0 (2019-10-26)

1. New Feature: Java 9, 10, 11, 12 support. But Java 6 and Java 7 are deprecated now.
1. New Feature: Log handler
1. New Feature: HTTP V2 support (thanks to Nginx v1.14.2)
1. Bug Fix: Memory leak with file handler #180
1. Bug Fix: Zero buffer error when hijack_send empty string #181
1. Bug Fix: Mysql driver issue about jdbc4 flag
1. Bug Fix: ContainsKey of nginx shared map
1. Bug Fix: NginxRequest.setVariable in a rewrite handler will hang
1. Bug Fix: Segmentation fault on shutdown
1. Bug Fix: Make clojure request map immutable for compojure
1. Enhancement: API for discarding request body (request.discardRequestBody())
1. Enhancement: Coroutine support for cascade constructor invoking
1. Enhancement: Configurable headers/variables prefetch for more safety in multithreaded mode 
1. Binaries Distribution: built with Nginx v1.14.2

## 0.4.5 (2017-05-28)

1. New Feature: Support to be compiled as Nginx dynamic module, thanks to [Andrew Hutchings](https://github.com/LinuxJedi)
1. Bug Fix: Cannot add multiple Cookies in a response
1. Bug Fix: Too many empty chunks are passed to Body filter & some body data lost
1. Enhancement: [Nginx-Jersey] Support jersey application sub class
1. Enhancement: Try to use enviroment variable JAVA_HOME to detect jvm when jvm_path is auto
1. Enhancement: NginxSharedHashMap.keySet/values/entrySet for debug/test usage.
1. Bug Fix: Can not use more than two shared maps.
1. Bug Fix: NullPointerExecption will happen when multiple rewrite handlers are invoked for one request
1. Bug Fix: Can't access ring request data in Sente handler. (content_handler_property fore-prefetch-all-properties true;)
1. Enhancement: Compile against Nginx 1.11 & Nginx 1.12
1. Bug Fix: Nginx reload will cause connection reset without response
1. Bug Fix: Header filter can not change response status from upstream
1. Bug Fix: body filters sometimes crash under thread pool mode
1. Binaries Distribution:  built with the latest stable Nginx v1.12.0 & openssl v1.1.0e

## 0.4.4 (2016-03-04)

1. New Feature: experimental nginx body filter by Java/Clojure/Groovy (issue #107)
1. New Feature: read request body by event callback (issue #109)
1. Bug Fix: 500 (internal server error) returns when committing 2000+ files to nginx as a proxy for apache mod_dav_svn (issue #106)

## 0.4.3 (2015-10-25)
1. New Feature: Add directive [jvm_classpath][] which supports wildcard character * (issue #95)
1. New Feature: Add directive [jvm_classpath_check][] which is enabled by default and when it is enabled access permission about classpaths will be checked.
1. New Feature: Add [NginxPubSubTopic(Java)/PubSubTopic(Clojure)][] to simplify handling messages among Nginx worker processes. (issue #97)
1. New Feature: [Shared Map based on shared memory][] (issue #96) and it has two implementations : tinymap & hashmap.
1. New Feature: [Shared Map based Ring session store][] (issue #98)
1. Enhancement: on-broadcast-event-decode!/on-broadcast! returns a removal function which can be used to remove the registered decoder/listener
1. Enhancement: embedded nginx-clojure becomes friendly to mock tests and  also fix issue #101
1. Bug Fix: After stopping an embedded Nginx-Clojure server keep-alived connections become CLOSE_WAIT.
1. Bug Fix: HackUtil.decode decodes unnecessary chars when several strings share one char[] generally on JDK 6
1. Bug Fix: jvm crashes with thread pool mode when open_file_cache is enabled.
1. Bug Fix: Fix compile errors when no sha1-implementation/zlib can be found (issue #99)
1. Example Project: Add [an example project about clojure web dev][] to show how to develop & deploy with Nginx-Clojure. Thanks to [Peter Taoussanis](https://github.com/ptaoussanis) without whose
comments there would not be such example project. (issue #91)
1. Documents: Add [Directives Reference](http://nginx-clojure.github.io/directives.html)


## 0.4.2 (2015-08-31)

1. New Feature: Support Sente (issue #87, see [this PR](https://github.com/ptaoussanis/sente/pull/160))
1. New Feature: Per-message Compression Extensions (PMCEs) for WebSocket (issue #88)

src/c/ngx_http_clojure_jvm.h
/*
 *  Copyright (C) Zhang,Yuexiang (xfeep)
 */

#ifndef NGX_HTTP_CLOJURE_JVM_H_
#define NGX_HTTP_CLOJURE_JVM_H_

#include <jni.h>

#if defined(_WIN32) || defined(WIN32)

#pragma warning (disable : 4305)
#pragma warning (disable : 4244)
#pragma warning (disable : 4152)


typedef jint (WINAPI *jni_createvm_pt)(JavaVM **pvm, void **penv, void *args);

#else

typedef jint (*jni_createvm_pt)(JavaVM **pvm, void **penv, void *args);

#endif


#define NGX_HTTP_CLOJURE_JVM_OK 0
#define NGX_HTTP_CLOJURE_JVM_ERR 1
#define NGX_HTTP_CLOJURE_JVM_ERR_LOAD_LIB 2
#define NGX_HTTP_CLOJURE_JVM_ERR_INIT_PIPE 3
#define NGX_HTTP_CLOJURE_JVM_ERR_MALLOC 4
#define NGX_HTTP_CLOJURE_JVM_ERR_INIT_MEMIDX 5
#define NGX_HTTP_CLOJURE_JVM_ERR_INIT_SOCKETAPI 6
#define NGX_HTTP_CLOJURE_JVM_ERR_INIT_SHAREDMAP 7

#define NGX_HTTP_CLOJURE_SOCKET_SHUTDOWN_READ 0
#define NGX_HTTP_CLOJURE_SOCKET_SHUTDOWN_WRITE 1
#define NGX_HTTP_CLOJURE_SOCKET_SHUTDOWN_BOTH 2
#define NGX_HTTP_CLOJURE_SOCKET_SHUTDOWN_SOFT_FLAG  4
#define NGX_HTTP_CLOJURE_SOCKET_SHUTDOWN_SOFT_READ  (NGX_HTTP_CLOJURE_SOCKET_SHUTDOWN_READ | NGX_HTTP_CLOJURE_SOCKET_SHUTDOWN_SOFT_FLAG)
#define NGX_HTTP_CLOJURE_SOCKET_SHUTDOWN_SOFT_WRITE (NGX_HTTP_CLOJURE_SOCKET_SHUTDOWN_WRITE | NGX_HTTP_CLOJURE_SOCKET_SHUTDOWN_SOFT_FLAG)
#define NGX_HTTP_CLOJURE_SOCKET_SHUTDOWN_SOFT_BOTH  (NGX_HTTP_CLOJURE_SOCKET_SHUTDOWN_BOTH | NGX_HTTP_CLOJURE_SOCKET_SHUTDOWN_SOFT_FLAG)


#define NGX_HTTP_CLOJURE_SOCKET_OK 0
#define NGX_HTTP_CLOJURE_SOCKET_ERR -16
#define NGX_HTTP_CLOJURE_SOCKET_ERR_RESOLVE -17
#define NGX_HTTP_CLOJURE_SOCKET_ERR_CONNECT -18
#define NGX_HTTP_CLOJURE_SOCKET_ERR_CONNECT_TIMEOUT -19
/**
 * @deprecated
 * please use either of NGX_HTTP_CLOJURE_SOCKET_ERR_CONNECT_TIMEOUT, NGX_HTTP_CLOJURE_SOCKET_ERR_READ_TIMEOUT or NGX_HTTP_CLOJURE_SOCKET_ERR_WRITE_TIMEOUT
 */
#define NGX_HTTP_CLOJURE_SOCKET_ERR_TIMEOUT -20
#define NGX_HTTP_CLOJURE_SOCKET_ERR_READ -21
#define NGX_HTTP_CLOJURE_SOCKET_ERR_READ_TIMEOUT -22
#define NGX_HTTP_CLOJURE_SOCKET_ERR_WRITE -23
#define NGX_HTTP_CLOJURE_SOCKET_ERR_WRITE_TIMEOUT -24
#define NGX_HTTP_CLOJURE_SOCKET_ERR_RESET -25
#define NGX_HTTP_CLOJURE_SOCKET_ERR_OUTOFMEMORY -26
#define NGX_HTTP_CLOJURE_SOCKET_ERR_AGAIN -27
#define NGX_HTTP_CLOJURE_SOCKET_ERR_BIND   -28


#define NGX_HTTP_CLOJURE_CHANNEL_EVENT_CLOSE 0
#define NGX_HTTP_CLOJURE_CHANNEL_EVENT_CONNECT  1
#define NGX_HTTP_CLOJURE_CHANNEL_EVENT_READ  2
#define NGX_HTTP_CLOJURE_CHANNEL_EVENT_WRITE 4
#define NGX_HTTP_CLOJURE_CHANNEL_EVENT_MSGREMAIN 8
#define NGX_HTTP_CLOJURE_CHANNEL_EVENT_MSGTEXT  16
#define NGX_HTTP_CLOJURE_CHANNEL_EVENT_MSGBIN   32
#define NGX_HTTP_CLOJURE_CHANNEL_EVENT_MSGCLOSE 64
#define NGX_HTTP_CLOJURE_CHANNEL_EVENT_MSGFIRST 128
#define NGX_HTTP_CLOJURE_CHANNEL_EVENT_MSGPONG 256


#define exception_handle(test, env, action) \
    do { \
        if (test) { \
			if ( (*env)->ExceptionOccurred(env) ) { \
				(*env)->ExceptionDescribe(env); \
				(*env)->ExceptionClear(env); \
				action; \
			} \
		} \
	} while (0)

#define ngx_http_clojure_abs_off_addr(jobj, off) \
		(jobj ? (void *)((char *)(*(uintptr_t*)jobj) + off) : (void *)(uintptr_t)off)

#define ngx_http_clojure_abs_addr(jobj) \
		(jobj ? (void *)((char *)(*(uintptr_t*)jobj)) : (void *)(uintptr_t)0;

extern int ngx_http_clojure_is_embeded_by_jse;

int ngx_http_clojure_check_jvm();

/*
 * initialize jvm and return NGX_HTTP_CLOJURE_JVM_OK for success or NGX_HTTP_CLOJURE_JVM_ERR* for failing.
 * if jvm has been successfully initialized already it will do nothing and just return NGX_HTTP_CLOJURE_JVM_OK
 */
int ngx_http_clojure_init_jvm(char *jvm_path, char * *opts, size_t len);

int ngx_http_clojure_init_by_jse_app(JNIEnv *env);

int ngx_http_clojure_get_env(JNIEnv **penv);

int ngx_http_clojure_get_jvm(JavaVM  **pvm);

int ngx_http_clojure_close_jvm();

#endif /* NGX_HTTP_CLOJURE_JVM_H_ */
